From 92ee842634222c1afc3b7fb4b3652f997a3435e7 Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Thu, 7 Aug 2014 07:06:19 -0400
Subject: [net] ah4: Use the IPsec protocol multiplexer API

Message-id: <1407395194-3809-4-git-send-email-jpirko@redhat.com>
Patchwork-id: 87517
O-Subject: [patch rhel7 net 03/18] ah4: Use the IPsec protocol multiplexer API
Bugzilla: 1091561
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

BZ1091561
https://brewweb.devel.redhat.com/taskinfo?taskID=7794261
Tested by jtluka's LNST smoke test recipe

upstream commit e5b56454e09a45ea6206d5253f78042c4e63f7d4
Author: Steffen Klassert <steffen.klassert@secunet.com>
Date:   Fri Feb 21 08:41:09 2014 +0100

    ah4: Use the IPsec protocol multiplexer API

    Switch ah4 to use the new IPsec protocol multiplexer.

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/ah4.c b/net/ipv4/ah4.c
index 2e7f194..044aaad 100644
--- a/net/ipv4/ah4.c
+++ b/net/ipv4/ah4.c
@@ -397,7 +397,7 @@ out:
  return err;
 }
 
-static void ah4_err(struct sk_buff *skb, u32 info)
+static int ah4_err(struct sk_buff *skb, u32 info)
 {
  struct net *net = dev_net(skb->dev);
  const struct iphdr *iph = (const struct iphdr *)skb->data;
@@ -407,17 +407,17 @@ static void ah4_err(struct sk_buff *skb, u32 info)
  switch (icmp_hdr(skb)->type) {
  case ICMP_DEST_UNREACH:
   if (icmp_hdr(skb)->code != ICMP_FRAG_NEEDED)
-   return;
+   return 0;
  case ICMP_REDIRECT:
   break;
  default:
-  return;
+  return 0;
  }
 
  x = xfrm_state_lookup(net, skb->mark, (const xfrm_address_t *)&iph->daddr,
          ah->spi, IPPROTO_AH, AF_INET);
  if (!x)
-  return;
+  return 0;
 
  if (icmp_hdr(skb)->type == ICMP_DEST_UNREACH) {
   atomic_inc(&flow_cache_genid);
@@ -427,6 +427,8 @@ static void ah4_err(struct sk_buff *skb, u32 info)
  } else
   ipv4_redirect(skb, net, 0, 0, IPPROTO_AH, 0);
  xfrm_state_put(x);
+
+ return 0;
 }
 
 static int ah_init_state(struct xfrm_state *x)
@@ -508,6 +510,10 @@ static void ah_destroy(struct xfrm_state *x)
  kfree(ahp);
 }
 
+static int ah4_rcv_cb(struct sk_buff *skb, int err)
+{
+ return 0;
+}
 
 static const struct xfrm_type ah_type =
 {
@@ -521,11 +527,12 @@ static const struct xfrm_type ah_type =
  .output  = ah_output
 };
 
-static const struct net_protocol ah4_protocol = {
+static struct xfrm4_protocol ah4_protocol = {
  .handler = xfrm4_rcv,
+ .input_handler = xfrm_input,
+ .cb_handler = ah4_rcv_cb,
  .err_handler = ah4_err,
- .no_policy = 1,
- .netns_ok = 1,
+ .priority = 0,
 };
 
 static int __init ah4_init(void)
@@ -534,7 +541,7 @@ static int __init ah4_init(void)
   pr_info("%s: can't add xfrm type\n", __func__);
   return -EAGAIN;
  }
- if (inet_add_protocol(&ah4_protocol, IPPROTO_AH) < 0) {
+ if (xfrm4_protocol_register(&ah4_protocol, IPPROTO_AH) < 0) {
   pr_info("%s: can't add protocol\n", __func__);
   xfrm_unregister_type(&ah_type, AF_INET);
   return -EAGAIN;
@@ -544,7 +551,7 @@ static int __init ah4_init(void)
 
 static void __exit ah4_fini(void)
 {
- if (inet_del_protocol(&ah4_protocol, IPPROTO_AH) < 0)
+ if (xfrm4_protocol_deregister(&ah4_protocol, IPPROTO_AH) < 0)
   pr_info("%s: can't remove protocol\n", __func__);
  if (xfrm_unregister_type(&ah_type, AF_INET) < 0)
   pr_info("%s: can't remove xfrm type\n", __func__);
-- 
1.7.1